torch/csrc/jit/passes/check_strict_fusion.cpp

```diff
@@ -0,0 +1,123 @@
+
+#include <torch/csrc/jit/passes/check_strict_fusion.h>
+
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/frontend/error_report.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/quantization/helper.h>
+#include <torch/csrc/jit/runtime/graph_iterator.h>
+#include <unordered_map>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+bool isStrictFusion(Value* value) {
+  const auto class_name = getModuleName(value);
+  return class_name.has_value() &&
+      (*class_name == "__torch__.torch.jit.strict_fusion");
+}
+
+} // namespace
+
+bool fusionGuardCheck(Symbol k) {
+  return k == Symbol::prim("TensorExprDynamicGuard") || k == prim::TypeCheck ||
+      k == prim::CudaFusionGuard || k == prim::RequiresGradCheck;
+}
+
+std::unordered_set<Node*> collectValuesUsedInGuard(
+    Node* guarding_if,
+    Node* enter_node) {
+  // DFS to collect
+  std::unordered_set<Node*> visited_nodes;
+  std::vector<Node*> queue = {guarding_if};
+
+  while (!queue.empty()) {
+    Node* curr = queue[queue.size() - 1];
+    queue.pop_back();
+    visited_nodes.insert(curr);
+    // these nodes directly test Tensor inputs, and are not part of additional
+    // guards inserted
+    if (fusionGuardCheck(curr->kind())) {
+      continue;
+    }
+    for (Value* v : curr->inputs()) {
+      Node* inp_node = v->node();
+      if (inp_node->isBefore(enter_node) ||
+          inp_node->owningBlock() != enter_node->owningBlock()) {
+        continue;
+      }
+      if (visited_nodes.count(inp_node)) {
+        continue;
+      }
+      queue.push_back(inp_node);
+    }
+  }
+  return visited_nodes;
+}
+
+void checkForUnfusedOps(Node* enter_node) {
+  std::vector<Node*> unsupported_nodes;
+  Node* guarding_if = nullptr;
```

Comment:
is it possible to have a case where there are two separate fusion groups without any other unsupported nodes in between?

Reply:
Not in the current fusers, but yea i'll add a check for that